chore(text-field): Split out input into subelement

bonniezhou opened the pull request:
Refactoring input logic into a subdirectory for readability.
Resolves #1595.

lynnjepsen commented on packages/mdc-textfield/adapter.js:
```diff
@@ -196,6 +175,12 @@ class MDCTextFieldAdapter {
+
+  /**
+   * Returns the foundation for the input element.
+   * @return {?MDCTextFieldInputFoundation}
```
I think the input foundation is actually required (unlike helper text, which was optional). So this should be `@return {MDCTextFieldInputFoundation}` (no ?)

kfranqueiro replied:
Usually in cases like this, we should have `!` instead of `?`, right? I mean, it shouldn't be null either?

bonniezhou replied:
Yeah, `{MDCTextFieldInputFoundation}` implies `{?MDCTextFieldInputFoundation}`, so I changed the `?` to `!`

lynnjepsen commented on packages/mdc-textfield/foundation.js:
```diff
@@ -91,16 +90,17 @@ class MDCTextFieldFoundation extends MDCFoundation {
     // Ensure label does not collide with any pre-filled value.
-    if (this.getNativeInput_().value) {
+    const input = this.adapter_.getInputFoundation();
+    if (input && input.getValue()) {
```
input is required. So you should only have to check input.getValue

lynnjepsen commented on packages/mdc-textfield/foundation.js:
```diff
@@ -128,11 +128,13 @@ class MDCTextFieldFoundation extends MDCFoundation {
-
-    this.receivedUserInput_ = true;
+    const input = this.adapter_.getInputFoundation();
+    if (input) {
```
input is required, so you don't need this if statement

lynnjepsen commented on packages/mdc-textfield/foundation.js:
```diff
@@ -244,23 +235,27 @@ class MDCTextFieldFoundation extends MDCFoundation {
   isDisabled() {
-    return this.getNativeInput_().disabled;
+    const input = this.adapter_.getInputFoundation();
+    return input ? input.isDisabled() : false;
```
input is required, so you don't have to check if it exists

lynnjepsen commented on packages/mdc-textfield/foundation.js:
```diff
@@ -244,23 +235,27 @@ class MDCTextFieldFoundation extends MDCFoundation {
     const {DISABLED, INVALID} = MDCTextFieldFoundation.cssClasses;
-    this.getNativeInput_().disabled = disabled;
+    const input = this.adapter_.getInputFoundation();
+    if (input) {
```
input is required, dont need the if statement

lynnjepsen commented on packages/mdc-textfield/index.js:
```diff
@@ -61,11 +63,17 @@ class MDCTextField extends MDCComponent {
+    bottomLineFactory = (el) => new MDCTextFieldBottomLine(el),
+    inputFactory = (el) => new MDCTextFieldInput(el)) {
+    const inputElement = this.root_.querySelector(strings.INPUT_SELECTOR);
+    if (inputElement) {
```
I think we should assume the input element exists

lynnjepsen commented on packages/mdc-textfield/index.js:
```diff
@@ -98,17 +106,12 @@ class MDCTextField extends MDCComponent {
     if (this.helperText_) {
       this.helperText_.destroy();
     }
+    if (this.input_) {
```
input is required, so you dont need the if statement

lynnjepsen commented on packages/mdc-textfield/index.js:
```diff
@@ -161,6 +164,16 @@ class MDCTextField extends MDCComponent {
       deregisterTextFieldInteractionHandler: (evtType, handler) => this.root_.removeEventListener(evtType, handler),
       notifyIconAction: () => this.emit(MDCTextFieldFoundation.strings.ICON_EVENT, {}),
+      registerInputEventHandler: (evtType, handler) => {
+        if (this.input_) {
```
dont need if statement

lynnjepsen commented on packages/mdc-textfield/index.js:
```diff
@@ -34,7 +36,7 @@ class MDCTextField extends MDCComponent {
   constructor(...args) {
     super(...args);
-    /** @private {?Element} */
+    /** @private {?MDCTextFieldInput} */
```
you can remove the ? here too

acdvorak replied:
I think we want to replace `?` with `!` since input is required

lynnjepsen commented on packages/mdc-textfield/input/README.md:
```diff
@@ -0,0 +1,67 @@
+
+<ul class="icon-list">
+  <li class="icon-list-item icon-list-item--spec">
+    <a href="https://material.io/guidelines/components/text-fields.html#text-fields-states">Material Design guidelines: Text Fields States</a>
```
I think this link

https://material.io/guidelines/components/text-fields.html#text-fields-field-types

might be a better place to link to

bonniezhou replied:
Done

bonniezhou commented:
Made input a required element and removed unnecessary checks for its existence. Also updated tests with a mock input, and added one to assert an error if the `mdc-text-field__input` element isn't found.

acdvorak commented:
LGTM!

**Caveat:** I have limited experience with our JS (I've mainly been in Sass land), so Lynn's review is much more trustworthy 😛

lynnjepsen commented on packages/mdc-textfield/input/README.md:
```diff
@@ -0,0 +1,66 @@
+
+##### MDCTextFieldInputFoundation.checkValidity() => boolean
+
+Returns the result of the checkValidity() call in the native input.
```
add a comment about what this returns if the client is using customized validity checking

lynnjepsen commented on packages/mdc-textfield/foundation.js:
```diff
@@ -128,11 +128,11 @@ class MDCTextFieldFoundation extends MDCFoundation {
-
-    this.receivedUserInput_ = true;
+    const input = this.adapter_.getInputFoundation();
+    input.setReceivedUserInput(true);
```
Can this setReceivedUserInput be handled internally by the MDCTextFieldInputFoundation?

bonniezhou replied:
Changed the input method to handleTextFieldInteraction()

kfranqueiro commented on packages/mdc-textfield/input/README.md:
```diff
@@ -0,0 +1,66 @@
+
+##### MDCTextFieldInputFoundation.checkValidity() => boolean
+
+Returns the result of the checkValidity() call in the native input. This can be the result of customized validity checking.
```
This isn't true, is it? The `checkValidity` foundation method simply results in calling `checkValidity` directly on the native input element. (Don't we already have an existing issue related to there currently being no way to retrieve custom validity set by `setValid`?)

lynnjepsen replied:
Ya I think you're right Ken. I guess we don't even reference custom validity in the input subelement foundation. We only reference custom validity in the parent MDCTextFieldFoundation. So we should drop this from the README.

Sorry for the confusion Bonnie, but revert this back to how it was.

kfranqueiro commented on packages/mdc-textfield/input/foundation.js:
```diff
@@ -0,0 +1,181 @@
+  /**
+   * @return {boolean} True if the input fails validity checks.
+   */
+  isBadInput() {
```
I realize this code existed prior to this split, but...I'm confused why this exists. What is this doing distinct from checkValidity? cc @amsheehan

lynnjepsen replied:
Lets not architecture in this PR. This PR is just refactoring. If it illuminates some problems with the old MDCTextFieldFoundation code...great! But file issues and we will resolve in another PR.

kfranqueiro commented on packages/mdc-textfield/input/foundation.js:
```diff
@@ -0,0 +1,181 @@
+   * @private
+   */
+  getNativeInput_() {
+    return this.adapter_.getNativeInput() ||
```
I realize this existed prior to the split, but...is there ever not an input element??

lynnjepsen replied:
Its annotated in the `MDCTextFieldInputAdapter` as `{?Element|?NativeInputType}`. So technically it is optional. But again, lets not architecture in this PR. This PR is just refactoring. If it illuminates some problems with the old MDCTextFieldFoundation code...great! But file issues and we will resolve in another PR.

kfranqueiro commented on packages/mdc-textfield/input/foundation.js:
```diff
@@ -0,0 +1,181 @@
+  /**
+   * @return {string} Returns the value in the input.
+   */
+  getValue() {
```
These APIs seem awkward for a foundation when they're just passing through a straight-up DOM operation. Is this something that would no longer be needed after #1675?

(I realize some of these e.g. isDisabled also existed prior to this PR...)

lynnjepsen replied:
The only other "non-awkward", option I can see is to have the foundation return `NativeInputType`. Like a `getNativeInputType()`. But I think I prefer we proxy the methods on `NativeInputType` as methods on the `MDCTextFieldInputFoundation` (aka .getValue and .isDisabled)

We would still need this after #1675. That would only change how we initiate these classes.
